R/generics.R

```diff
@@ -102,8 +102,27 @@ dplyr_col_modify <- function(data, cols) {
 
 #' @export
 dplyr_col_modify.data.frame <- function(data, cols) {
-  data[names(cols)] <- cols
-  data
+  # Implement from first principles to avoiding edge cases in [.data.frame
+  # and [.tibble in 2.1.3 and earlier
+
+  # Apply tidyverse recycling rules
+  cols <- lapply(cols, vec_recycle, size = nrow(data))
+
+  names(cols) <- as_utf8_character(names2(cols))
+  names(data) <- as_utf8_character(names2(data))
+
+  out <- vec_data(data)
+  out[names(cols)] <- cols
+
+  is_null <- map_lgl(out, is.null)
+  out <- out[!is_null]
```

Comment:
I'd really like to avoid iterating over `out`, surprisingly this might be the fastest solution in many cases. Should we do this only if there are `NULL` values in `cols`?

```suggestion
  # Iterating over out will be faster than splitting cols into
  # NULL and non-NULL values in many cases.
  if (any(map_lgl(cols, is.null))) {
    is_null <- map_lgl(out, is.null)
    out <- out[!is_null]
  }
```

Reply:
We could also do a for loop + `[[`:

```R
for (col in seq_along(cols)) {
  nm <- names(cols)[[i]]
  data[[nm]] <- cols[[i]]
```

Reply:
That's how tibble is doing it.